### ISSUE TITLE
Package management system

### ISSUE THREAD

**Issue by quentincaffeino** [godotengine/godot#17092]
Hello,

Lately I'm thinking about adding some package management system to godo more often as I move bunch of my libs from project to project.
As I think it would not be that hard to take any of highly used program (such as composer, yarn, pip etc.) and adapting it principles for godot. Also as there's already an asset lib website, it is enough to update it to support keeping packages database.
Also a name have come to my mind: GVen [Gwen], Godot Vendor.

What do you think?

Best :smiley:

**Comment by groud** [comment]
Honestly, I don't see the point. The only thing I can see as useful would be the dependency manager.
(having a command-line tool would be useless since we already have the editor)
However, assets are rarely dependent on each other, so for me the asset store is enough. But we could add dependency management to the assets store I guess.

**Comment by avril-gh** [comment]
We have very nice "asset library" function out there.
you can share your libs, assets, plugins, games, whatever with community or keep it local.

**Comment by QuentinCaffeino** [comment]
Firstly also sorry for my great English.

> having a command-line tool would be useless since we already have the editor

Of course first we should adapt editor for it. But having a cli is also a good option (or at least having it integrated in godot executable so it can be used through parameters)


> However, assets are rarely dependent on each other

And that sometimes is a bad thing because you write same pieces of code again and again instead of making maximum profit of code you or someone else wrote before and spending your time on some other more important tasks. 


> But we could add dependency management to the assets store I guess.

Well yeah, that's the main point. To be able to create and share reusable pieces of code, or "packages".
But it isn't only about adding dependency column to assets db. It requires a bit more work than that.


What I propose is to make a standard. It is required for several reasons:
1. Any package can be installed via editor or cli using just a package name.
2. Then any package could install it dependencies along with it using also just package names.
3. To be able to use each other they need some sort of structure or "namespaces" defined by a standart. And also same structure is used so you can use this packages.

IMO it is better to take a composer as an example for creating such thing in godo. So I suggest you take a look at it.

**Comment by avril-gh** [comment]
I know its kinda too long to read, but maybe someone would care to check sometime how something looks from another point of view.

**My 3 cents about composer and "things that claims to make life more easy"**

I been contributing a lot to myBB forum software in the past. I been very passionate about it and spend a lot of time working on this project. If i could continue it, i could be today "staff member" like everyone else are now, with whom we worked together.

There been plans to make myBB2 which has to be more rewriting from scratch than next milestone and team decided that its good oportunity to add stuff like composer and many third party frameworks that has to make "life easier" and everything "more shiny".

However, it ended up being not only different piece of software, but also all those composers and "trendy" frameworks simply added alot of extra hassle.

Because of this i stop contibuting to myBB. As a developer, personaly i see a big internal potency of (old) myBB that goes far beyound normal forum software, so im still observing it.
Unfortunately i noticed, from that time, releases count has fallen dramaticaly, Threses very low count of contributors (myBB2 have only 20 contributors and one merge per month or more) and it seems that project suffer big stalls, (dont want to say its dying)
So, it seems like things that has to "made it easy" for a fiew, for others added extra unnecessary layers of code increasing complexity and additional hassle.

**More about things that have to "make life easier"** 
Since its important for me to reach widest market possible, i aim at GLES2, that meant, i use Godot 2.x and observe with passion, high expectations and hopes, whats goin on in 3.x branch with hope it'll be great. But there are some fears i allready experienced in 3.x

When i tested it last time and tried to remake my 2.x projects to see how it "feels" on 3.x,
i found that its kinda hard to make clean project hierarchy. I meant well defined folders which contain related stuff. Like here game levels, here assets, here code, here textures, there sound ect ect.
Theres many things like "you cant change this or that because of something" where are textures and why not where i want them to be, and why they auto reimport (overwrites my changes) and so on and so on,
Not only hard to figure out how to nicely create project hierarchy (folders with data) so i can find everything fast and reuse it as much as possible to save resources, but also there is new folder .import which looks like one big trashcan with everything droped one onto another with names like bc66xc5445vxc65v44bvcb4567654gdfgdfg that is allready scary !

I do not assume nor judge anything. I tried 3.x - not used it yet serious way so it might be my lack of knowledge and experience with 3.x
Just the short test kinda worried me if its going in good way and if something that has to be "easy" not turned into additional hassle and now i will have to fight Godot to keep clean project stucture and reuse resources in the clear manageable way against how it want to do it (store resources in unknown locations, with cryptic names, use it "somehow" reuse ? ect ect)
Again. i might be wrong (i hope i am) since i have no serious experience in using 3.x yet. Just worried.

**Packaging topic**
I seen here and there topics about "lets make Godot installers" ect, it'll make life easy - click and go.
For the one who made it - yes it is. Im doing many "one click" things for myself to speed up work,
However for others its - one click ~ and... what happened ?
When i get something from repo i see what is it, i can compile it, i can unzip something, i see it - i understand it. To understand (whatever it could be) is most important for us all.

When i have automated installer, i click and i dont know whats going on.
What been inside and where it went, what it done to my computer, what it changed, does it installed / changed something i dont want to be changed ? If there been some additional stuff i didnt asked for but i got it installed (like very cool and important auto updaters which makes my life even more easy, and scan my computer then send detailed info on software im runing, what time i have computer turned on and what i eat on lunch to guys in china who will analyze it and "help me")
Ofcourse i can read documentation but its been promised to easy and now i must do additional work - figure out what exacly happened to my computer.

...look i just unzip, look whats inside, and put it where i need it - thats easy.

**Comment by QuentinCaffeino** [comment]
There are always two sides of medal. My propose is about adding more structured access to modules on asset store (they still will be on github actually but store will carry some data so it is easy to find those modules). But no one makes you or any body else use it. You can still get zip from github and use it as you want. Same with composer. Composer is just a package manager. How and if you use it is on you. So IMO myBB team haven't thought enough about why do they need composer and what packages do they need.

**Comment by avril-gh** [comment]
I didnt say, yes or no, just sharing thoughts.
What i meant is - Striping from unnecessary mechanical repeatable work seems good, but striping developer from thinking might not be good.

**Comment by groud** [comment]
What you don't get is that, in video games, there's not a single manner way of doing things.
Basically, when you download some assets on the assets store, you download something that you can modify and adapt to your own needs, while you rarely need something that has a given version and shouldn't be modified. That's why in most cases, having a package manager makes no sense, as in the end, everyone is going to have a modified version of the resource. (Dependencies are going  broken if you modify the resource)

The only place this seems to have a meaning is basically for plugins, which, IMHO, does not need such a complex package manager. Probably something like blender does should be good enough.

**Comment by avril-gh** [comment]
like @groud mentioned "(...) there's not a single manner way of doing things (...)" and i could say the same in my own words "theres as many truths and ways of doing something as count of humans" were drifting on a very wide topic.

Could be sensible if @QuentinCaffeino would like to introduce to us more detailed draft of his idea, like where he want it to implemented, what options could be available there, and why its good for the most of us, so we could discuss more to the topic.

**Comment by vnen** [comment]
I understand the angle to propose this, but game development, as said before, works in a different way. Package managers are common in projects that live a long time and are constantly updated. This kind of software is also usually made by a large team of programmers which come and go, so the setup happens many times during the lifetime of the product.

Games are more like a "one-off" product. You make it, release it, and then forget it. Well, of course you release bugfixes, but updating the dependencies is only made if it's really needed, because it can introduce more bugs. It's also common to have a fixed team that goes in the project from beginning to end, so the setup usually happens only once.

Of course, in the end the community kind of dictates what's available or not. If you decide to make it and everybody starts using it, then we can consider making it official. But for now I don't see a need for an automatized package manager.

**Comment by herman-rogers** [comment]
Something that's not said that is fairly obvious is that usually when downloading assets you save them in a folder that is then versioned and stored on git. In my opinion this is pretty bad - you have a massive assets folder that then gets copied across the team and you have the overhead of adding to the mess. Perhaps its easy to manage a small project when its only you (or a game jam). But for a production product this can become really cumbersome.

**Comment by elopezga** [comment]
@herman-rogers I agree with this sentiment. There is value in developing assets independently from your project to increase modularity and reusability. Having your assets in separate repos introduces maintainability problems if you simply import them into the project and store them with git, as you mentioned.

In web development, tools such as NPM are very popular because you can include packages into your project without including them in your git repo; you only include the configuration file that tells NPM what packages to include. You benefit from such a system in two ways: everyone in your project automatically has the same dependencies as you (versions and all), and you can easily implement build automation on the cloud. I think having a package manager is great for small projects, and essential for large projects.

The software industry seems to be heading into this type of thinking when dealing with assets. For example, Unity has recently introduced their own package manager to replace their current asset acquisition method. Furthermore, it opens up a platform for people to be able to easily share their assets across projects and with each other. With an abundance of assets to choose from, and the ability to easily import these assets into your project, we can see a massive increase in productivity for our own projects. I believe Godot can greatly benefit from having a package manager.

**Comment by Calinou** [comment]
Closing in favor of https://github.com/godotengine/godot-proposals/issues/142 and https://github.com/godotengine/godot-proposals/issues/607, as feature proposals are now tracked on the Godot proposals repository.
